test/org/zaproxy/zap/spider/parser/SpiderHtmlFormParserUnitTest.java

```diff
@@ -0,0 +1,243 @@
+        // Given
+        SpiderHtmlFormParser htmlParser = createSpiderHtmlFormParser();
+        boolean parsed = true;
+        // When
+        boolean canParse = htmlParser.canParseResource(new HttpMessage(), ROOT_PATH, parsed);
+        // Then
+        assertThat(canParse, is(equalTo(false)));
+    }
+
+    @Test
+    public void shouldNotParseNonHtmlResponse() {
+        // Given
+        SpiderHtmlFormParser htmlParser = createSpiderHtmlFormParser();
+        boolean parsed = false;
+        // When
+        boolean canParse = htmlParser.canParseResource(new HttpMessage(), ROOT_PATH, parsed);
+        // Then
+        assertThat(canParse, is(equalTo(false)));
+    }
+
+    @Test
+    public void shouldParseHtmlResponse() {
+        // Given
+        SpiderHtmlFormParser htmlParser = createSpiderHtmlFormParser();
+        HttpMessage messageHtmlResponse = createMessageWith("NoForms.html");
+        boolean parsed = false;
+        // When
+        boolean canParse = htmlParser.canParseResource(messageHtmlResponse, ROOT_PATH, parsed);
+        // Then
+        assertThat(canParse, is(equalTo(true)));
+    }
+
+    @Test
+    public void shouldParseHtmlResponseEvenIfProvidedPathIsNull() {
+        // Given
+        SpiderHtmlFormParser htmlParser = createSpiderHtmlFormParser();
+        HttpMessage messageHtmlResponse = createMessageWith("NoForms.html");
+        boolean parsed = false;
+        // When
+        boolean canParse = htmlParser.canParseResource(messageHtmlResponse, null, parsed);
+        // Then
+        assertThat(canParse, is(equalTo(true)));
+    }
+
+    @Test
+    public void shouldNotParseHtmlResponseIfAlreadyParsed() {
+        // Given
+        SpiderHtmlFormParser htmlParser = createSpiderHtmlFormParser();
+        HttpMessage messageHtmlResponse = createMessageWith("NoForms.html");
+        boolean parsed = true;
+        // When
+        boolean canParse = htmlParser.canParseResource(messageHtmlResponse, ROOT_PATH, parsed);
+        // Then
+        assertThat(canParse, is(equalTo(false)));
+    }
+
+    @Test(expected = NullPointerException.class)
+    public void shouldFailToParseAnUndefinedMessage() {
+        // Given
+        HttpMessage undefinedMessage = null;
+        SpiderHtmlFormParser htmlParser = createSpiderHtmlFormParser();
+        Source source = createSource(createMessageWith("NoForms.html"));
+        // When
+        htmlParser.parseResource(undefinedMessage, source, BASE_DEPTH);
+        // Then = NullPointerException
+    }
+
+    @Test
+    public void shouldParseMessageEvenWithoutSource() {
+        // Given
+        Source source = null;
+        SpiderHtmlFormParser htmlParser = createSpiderHtmlFormParser();
+        HttpMessage messageHtmlResponse = createMessageWith("NoForms.html");
+        // When
+        htmlParser.parseResource(messageHtmlResponse, source, BASE_DEPTH);
+        // Then = No exception
+    }
+
+    @Test
+    public void shouldNeverConsiderCompletelyParsed() {
```

Comment:
Whether or not it's considered parsed does not depend on the inputs, that's why it just says that. Why it has that behaviour? I guess because it's a fall back parser.